### PR TITLE
8300201: When storing MemoryAddress.ofLong(0x0000000080000000L), MemorySegment.get is not equal to MemorySegment.set because of the expanded sign

### PR DESCRIPTION
**Pull request by minborg**
This PR proposes to normalize the upper 32-bit to zero on 32-bit systems.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8300201](https://bugs.openjdk.org/browse/JDK-8300201): When storing MemoryAddress.ofLong(0x0000000080000000L), MemorySegment.get is not equal to MemorySegment.set because of the expanded sign

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/774/head:pull/774` \
`$ git checkout pull/774`

Update a local copy of the PR: \
`$ git checkout pull/774` \
`$ git pull https://git.openjdk.org/panama-foreign pull/774/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 774`

View PR using the GUI difftool: \
`$ git pr show -t 774`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/774.diff">https://git.openjdk.org/panama-foreign/pull/774.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 1]
```diff
@@ -1042,6 +1042,11 @@ static MemorySegment ofArray(double[] doubleArray) {
      * {@snippet lang = java:
      * ofAddress(address, 0);
      *}
+     * <p>
+     * On 32-bit platforms, the given address value will be normalized such that the
+     * returned memory segment's method {@link MemorySegment#address() address} will always return
+     * zeros for the upper 32 bits of the address.
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1048]:
```suggestion
     * On 32-bit platforms, the given address value will be normalized such that the
     * upper 32 bits of the {@link MemorySegment#address() address} of returned memory segment
     * are set to zero.
```

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 2]
```diff
@@ -1044,8 +1044,8 @@ static MemorySegment ofArray(double[] doubleArray) {
      *}
      * <p>
      * On 32-bit platforms, the given address value will be normalized such that the
-     * returned memory segment's method {@link MemorySegment#address() address} will always return
-     * zeros for the upper 32 bits of the address.
+     * upper 32 bits of the {@link MemorySegment#address() address} of returned memory segment
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1047]: Sorry, I mean to replicate this change in all the other pieces of javadoc where you talk of 32 bits.

Comment (minborg) [reply src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1047]: Fixed now

**Comment by GavinRay97** [comment]
Sorry to follow on here, but a question related to this -- you can use the upper 16 bits of a 64-bit pointer for pointer tagging, and also the lower 3-bits if it's on a word-aligned boundary.

This is really useful for building things like atomic tagged pointers, where you can store an ABA-like counter in the 16 bits and use the 3 bits for whatever userdata tag you want:

```
// A word-aligned, atomic tagged pointer.
// Uses both the upper 16 bits for storage, and the lower 3 bits for tagging.
//
//   64                48                32                16
// 0xXXXXXXXXXXXXXXXX  0000000000000000  0000000000000000  0000000000000XXX
//   ^                 ^                                                ^
//   |                 |                                                +-- Tag (3 bits)
//   |                 +-- Pointer (48 bits)
//   +-- Counter (16 bits)
```

Is this possible with `MemoryAddress` and bitmasks currently in Java 🤔

**Comment by minborg** [comment]
> Sorry to follow on here, but a question related to this -- you can use the upper 16 bits of a 64-bit pointer for pointer tagging, and also the lower 3-bits if it's on a word-aligned boundary.
> 
> This is really useful for building things like atomic tagged pointers, where you can store an ABA-like counter in the 16 bits and use the 3 bits for whatever userdata tag you want:
> 
> ```
> // A word-aligned, atomic tagged pointer.
> // Uses both the upper 16 bits for storage, and the lower 3 bits for tagging.
> //
> //   64                48                32                16
> // 0xXXXXXXXXXXXXXXXX  0000000000000000  0000000000000000  0000000000000XXX
> //   ^                 ^                                                ^
> //   |                 |                                                +-- Tag (3 bits)
> //   |                 +-- Pointer (48 bits)
> //   +-- Counter (16 bits)
> ```
> 
> Is this possible with `MemoryAddress` and bitmasks currently in Java 🤔
> 
> (Or does an optimization like that even make sense on the JVM?)

On some systems, mapped memory segments appear as negative longs and thus, is at least using bit 63.

**Comment by mcimadamore** [comment]
> Sorry to follow on here, but a question related to this -- you can use the upper 16 bits of a 64-bit pointer for pointer tagging, and also the lower 3-bits if it's on a word-aligned boundary.
> 
> This is really useful for building things like atomic tagged pointers, where you can store an ABA-like counter in the 16 bits and use the 3 bits for whatever userdata tag you want:
> 
> ```
> // A word-aligned, atomic tagged pointer.
> // Uses both the upper 16 bits for storage, and the lower 3 bits for tagging.
> //
> //   64                48                32                16
> // 0xXXXXXXXXXXXXXXXX  0000000000000000  0000000000000000  0000000000000XXX
> //   ^                 ^                                                ^
> //   |                 |                                                +-- Tag (3 bits)
> //   |                 +-- Pointer (48 bits)
> //   +-- Counter (16 bits)
> ```
> 
> Is this possible with `MemoryAddress` and bitmasks currently in Java thinking
> 
> (Or does an optimization like that even make sense on the JVM?)

On 64 bit platforms you can decide how to create your 64-bit addresses with `MemorySegment.ofAddress`. So, if you want to encode extra bit patterns on the high order bits, you can do so. But then you are also responsible for filtering these out, which require some extra case (because you have a memory segment whose base address is essentially invalid and cannot be used for access operation). Probably you will need/want to encapsulate the logic that deals with this segments. To prevent users from accessing these directly (which will likely lead to a crash).

That said, this has nothing to do with this PR, which is about when the user creates a new segment from a long address on a 32-bit system.

**Review by JornVernee — APPROVED** [review]

File: src/java.base/share/classes/java/lang/foreign/MemorySegment.java [hunk 3]
```diff
@@ -1068,8 +1068,8 @@ static MemorySegment ofAddress(long address) {
      * restricted methods, and use safe and supported functionalities, where possible.
      * <p>
      * On 32-bit platforms, the given address value will be normalized such that the
-     * returned memory segment's method {@link MemorySegment#address() address} will always return
-     * zeros for the upper 32 bits of the address.
+     * upper 32 bits of the {@link MemorySegment#address() address} of returned memory segment
```
Comment (mcimadamore) [inline src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1071]: Ah - late to the party. Is "upper bits" correct here? E.g. doesn't it depend on endianness?

Comment (mcimadamore) [reply src/java.base/share/classes/java/lang/foreign/MemorySegment.java:1071]: I'm partly confusing myself, but we should still draw from javadoc in other areas:
https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/lang/Long.html#highestOneBit(long)
This seems to use highest-order/leftmost for "uppermost bit" :-)

**Pull request closed without merging**